### PR TITLE
feat: distinguish HTTP status from Lassie status

### PR DESCRIPTION
**Pull request by bajtos**
Remove the following result codes:
- `ERROR_${number}`
- `BAD_GATEWAY`
- `GATEWAY_TIMEOUT`

Replace them with HTTP status codes prefixed with `HTTP_` if the
retrieval was performed directly over Trustless HTTP GW protocol
and `LASSIE_` if the retrieval was performed using Lassie. In the latter
case, the status code comes from Lassie, not the storage provider.

### CODE REVIEW

File: lib/preprocess.js [hunk 1]
```diff
@@ -281,7 +279,11 @@ export const getRetrievalResult = measurement => {
     case 904: return 'CANNOT_PARSE_CAR_FILE'
   }
 
-  if (measurement.status_code >= 300) return `ERROR_${measurement.status_code}`
+  if (measurement.status_code >= 300) {
+    const prefix = measurement.protocol === 'http' ? 'HTTP_' : 'LASSIE_'
+    // I cannot use + concatenation because TypeScript would complain
```
Comment (juliangruber) [inline lib/preprocess.js:284]:
```suggestion
```

String interpolation is the idiomatic JS way now anyway

File: lib/preprocess.js [hunk 2]
```diff
@@ -281,7 +279,11 @@ export const getRetrievalResult = measurement => {
...
+    return `${prefix}${measurement.status_code}`
```
Comment (juliangruber) [inline lib/preprocess.js:285]:
```suggestion
    return `${prefix}ERROR_${measurement.status_code}`
```

Shall we keep `ERROR_` to clarify the meaning? Or is this always an error anyway?

File: lib/retrieval-stats.js [hunk 3]
```diff
@@ -39,10 +39,7 @@ export const buildRetrievalStats = (measurements, telemetryPoint) => {
   /** @type {Partial<Record<import('./typings.js').RetrievalResult, number>>} */
   const resultBreakdown = {
     OK: 0,
-    TIMEOUT: 0,
-    CAR_TOO_LARGE: 0,
-    BAD_GATEWAY: 0,
-    GATEWAY_TIMEOUT: 0
```
Comment (juliangruber) [inline lib/retrieval-stats.js:45]: why does this PR remove these 3 result codes from this test?

Comment (juliangruber) [reply lib/retrieval-stats.js:45]: Shouldn't there be new `LASSIE_` or `HTTP_` codes?

File: test/helpers/test-data.js [hunk 4]
```diff
@@ -20,7 +20,7 @@ export const VALID_MEASUREMENT = {
   provider_address: '/dns4/production-ipfs-peer.pinata.cloud/tcp/3000/ws/p2p/Qma8ddFEQWEU8ijWvdxXm3nxU7oHsRtCykAaVz8WUYhiKn',
   providerId: 'PROVIDERID',
   spark_version: '1.10.4',
-  protocol: 'bitswap',
+  protocol: 'http',
```
Comment (juliangruber) [inline test/helpers/test-data.js:23]: what is this change for?

**Review by juliangruber — CHANGES_REQUESTED** [review]
+1 on the direction!

File: lib/preprocess.js (hunk 2 above)
Comment (bajtos) [reply lib/preprocess.js:285]: I guess it depends on how you define an error.

This function returns a "retrieval result code".

There is only one code for a retrieval that was successful - `OK`.

All other codes describe various failure modes.

That's why I think we don't need to include `ERROR` in those codes.

File: test/helpers/test-data.js (hunk 4 above)
Comment (bajtos) [reply test/helpers/test-data.js:23]: It's a bit unrelated. I thought this change would reduce the number of tests I need to fix as part of this PR, but after I reverted the change, all tests are passing, so it's not necessary at all.

Reverted.

File: lib/retrieval-stats.js (hunk 3 above)
Comment (bajtos) [reply lib/retrieval-stats.js:45]: See the comment above this object initialisation (GitHub diff does not show it by default, unfortunately):

```js
  // We are intentionally not initializing all possible keys here.
  // Example of omitted keys: UNDEFINED, ERROR_500 and ERROR_404.
  // The idea is that if we don't explicitly initialise them here and there is no measurement with
  // such retrieval result, then the Grafana dashboard will not show these results at all.
```

IIRC, I added a subset of keys to ensure they always show at the top of the legend in Grafana, and that we show these keys (retrieval result codes) even if no measurement reported the key.

Screenshot 2024-10-23 at 13 31 05

I feel this is no longer needed/useful for all of these codes.

On the second thought, maybe `TIMEOUT` is worth keeping.

WDYT?

Comment (bajtos) [reply lib/retrieval-stats.js:45]: >ensure they always show at the top of the legend in Grafana

This is no longer relevant. We have a transformation in Grafana to (re)order the codes manually.

Comment (juliangruber) [reply lib/retrieval-stats.js:45]: got it!

**Review by juliangruber — APPROVED** [review]

**Pull request merged**